Compiler crash: cannot convert struct to struct

### Describe the bug

Compiler crashed with message

error: cannot convert 'struct ...' to 'struct ...'


### Expected Behavior

Compiler should not crash.

### Current Behavior

```
---- Testing... ------------------------------------------------------------------------------------------------
----------------------------------------------------------------------------------------------------------------
==================                                                                                              
/tmp/.../run_test. ... .tmp.c:3892: error: cannot convert 'struct main__MyStruct' to 'struct _option_main__MyStruct'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```


### Reproduction Steps

run_test.v 
```

module main

struct MyStruct {
    field int
}

pub type MyOpt = ?MyStruct

fn empty() map[string]MyOpt {
  return {}
}

fn test_empty() {
	expected := {'key': MyOpt(none)}
	assert empty() == expected
}

```


```
$ v -stats test run_test.v 
```


### Possible Solution

_No response_

### Additional Information/Context

I was trying to determine if the `map` value for a given key can be `none`.

### V version

Current V version: V 0.3.4 acfe785

### Environment details (OS name and version, etc.)

OS: linux, Ubuntu 22.04.2 LTS

--- a/vlib/v/gen/c/auto_eq_methods.v
+++ b/vlib/v/gen/c/auto_eq_methods.v
@@ -230,7 +230,7 @@ fn (mut g Gen) gen_alias_equality_fn(left_type ast.Type) string {
 	mut fn_builder := strings.new_builder(512)
 	fn_builder.writeln('static bool ${ptr_styp}_alias_eq(${ptr_styp} a, ${ptr_styp} b) {')
 
-	is_option := info.parent_type.has_flag(.option) || left.typ.has_flag(.option)
+	is_option := left.typ.has_flag(.option)
 
 	left_var := if is_option { '*' + g.read_opt(info.parent_type, 'a') } else { 'a' }
 	right_var := if is_option { '*' + g.read_opt(info.parent_type, 'b') } else { 'b' }

--- a/vlib/v/gen/c/auto_eq_methods.v
+++ b/vlib/v/gen/c/auto_eq_methods.v
@@ -232,12 +232,14 @@ fn (mut g Gen) gen_alias_equality_fn(left_type ast.Type) string {
 
 	is_option := left.typ.has_flag(.option)
 
-	left_var := if is_option { '*' + g.read_opt(info.parent_type, 'a') } else { 'a' }
-	right_var := if is_option { '*' + g.read_opt(info.parent_type, 'b') } else { 'b' }
+	mut left_var := if is_option { '*' + g.read_opt(info.parent_type, 'a') } else { 'a' }
+	mut right_var := if is_option { '*' + g.read_opt(info.parent_type, 'b') } else { 'b' }
 
 	sym := g.table.sym(info.parent_type)
 	if sym.kind == .string {
-		if is_option {
+		if info.parent_type.has_flag(.option) {
+			left_var = '*' + g.read_opt(info.parent_type, 'a')
+			right_var = '*' + g.read_opt(info.parent_type, 'b')
 			fn_builder.writeln('\treturn ((${left_var}).len == (${right_var}).len && (${left_var}).len == 0) || string__eq(${left_var}, ${right_var});')
 		} else {
 			fn_builder.writeln('\treturn string__eq(a, b);')

--- a/vlib/v/gen/c/auto_eq_methods.v
+++ b/vlib/v/gen/c/auto_eq_methods.v
@@ -464,7 +464,11 @@ fn (mut g Gen) gen_map_equality_fn(left_type ast.Type) string {
 			fn_builder.writeln('\t\tif (*(voidptr*)map_get(&b, k, &(voidptr[]){ 0 }) != v) {')
 		}
 		else {
-			fn_builder.writeln('\t\tif (*(${ptr_value_styp}*)map_get(&b, k, &(${ptr_value_styp}[]){ 0 }) != v) {')
+			if value.typ.has_flag(.option) {
+				fn_builder.writeln('\t\tif (memcmp(v.data, ((${ptr_value_styp}*)map_get(&b, k, &(${ptr_value_styp}[]){ 0 }))->data, sizeof(${g.base_type(value.typ)})) != 0) {')
+			} else {
+				fn_builder.writeln('\t\tif (*(${ptr_value_styp}*)map_get(&b, k, &(${ptr_value_styp}[]){ 0 }) != v) {')
+			}
 		}
 	}
 	fn_builder.writeln('\t\t\treturn false;')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3923,7 +3923,7 @@ fn (mut g Gen) unlock_locks() {
 
 fn (mut g Gen) map_init(node ast.MapInit) {
 	unwrap_key_typ := g.unwrap_generic(node.key_type)
-	unwrap_val_typ := g.unwrap_generic(node.value_type).clear_flags(.option, .result)
+	unwrap_val_typ := g.unwrap_generic(node.value_type).clear_flags(.result)
 	key_typ_str := g.typ(unwrap_key_typ)
 	value_typ_str := g.typ(unwrap_val_typ)
 	value_sym := g.table.sym(unwrap_val_typ)

--- a/vlib/v/tests/option_alias_eq_test.v
+++ b/vlib/v/tests/option_alias_eq_test.v
@@ -10,9 +10,9 @@ fn empty() map[string]MyOpt {
 
 fn test_empty() {
 	expected := {
-		'key': MyOpt(none)
+		'key': ?MyOpt(none)
 	}
 	assert dump(expected) == {
-		'key': MyOpt(none)
+		'key': ?MyOpt(none)
 	}
 }

--- a/vlib/v/tests/option_alias_eq_test.v
+++ b/vlib/v/tests/option_alias_eq_test.v
@@ -1,0 +1,18 @@
+struct MyStruct {
+	field int
+}
+
+pub type MyOpt = ?MyStruct
+
+fn empty() map[string]MyOpt {
+	return {}
+}
+
+fn test_empty() {
+	expected := {
+		'key': MyOpt(none)
+	}
+	assert dump(expected) == {
+		'key': MyOpt(none)
+	}
+}

